pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go

```diff
@@ -116,40 +115,46 @@ type RaftInterface interface {
 	//
 	// If the peer is not in StateReplicate, this call does nothing.
 	SendPingRaftMuLocked(roachpb.ReplicaID) bool
-	// MakeMsgAppRaftMuLocked is used to construct a MsgApp for entries in
-	// [start, end) and must only be called in MsgAppPull mode for followers.
+	// SendMsgAppRaftMuLocked is used to construct a MsgApp for entries in the
+	// slice and must only be called in MsgAppPull mode for followers. Say
+	// [start, end) represent the entries in the slice.
 	//
-	// REQUIRES:
+	// REQUIRES (to the best knowledge of the caller):
+	// - start < end, i.e., the slice is non-empty.
 	// - replicaID i, is in StateReplicate.
 	// - start == Next(i)
 	// - end <= NextUnstableIndex
-	// - maxSize > 0.
+	//
+	// Returns false if a message cannot be generated. This could be because the
+	// knowledge of the caller is incorrect (which can happen because it is
+	// stale). See RawNode.SendMsgApp for the error conditions.
+	//
+	// If it returns true, all the entries in the slice are in the message, and
+	// Next is advanced to be equal to end.
+	SendMsgAppRaftMuLocked(replicaID roachpb.ReplicaID, slice RaftLogSlice) (raftpb.Message, bool)
+}
+
+// RaftLogSnapshot abstract raft.LogSnapshot.
+type RaftLogSnapshot interface {
+	// LogSlice returns a slice containing a prefix of [start, end). It must
+	// only be called in MsgAppPull mode for followers. The maxSize is required
+	// to be > 0.
 	//
 	// If the sum of all entries in [start,end) are <= maxSize, all will be
 	// returned. Else, entries will be returned until, and including, the first
 	// entry that causes maxSize to be equaled or exceeded. This implies at
-	// least one entry will be returned in the MsgApp on success.
-	//
-	// Returns an error if log truncated, or there is some other transient
-	// problem. If no error, there is at least one entry in the message, and
-	// Next is advanced to be equal to the index+1 of the last entry in the
-	// returned message.
-	//
-	// Requires Replica.raftMu to be held.
+	// least one entry will be returned in the slice on success.
 	//
-	// TODO(pav-kv): There are some rare non log truncation cases, where the
-	// flow stays in StateReplicate. We should define or eliminate these cases.
+	// Returns an error if the log is truncated, or there is some other
+	// transient problem.
 	//
-	// TODO(sumeer): This is a temporary API. LogSnapshot and LogSlice will
-	// replace it, and we will do this in two steps: (a) create a LogSlice while
-	// holding raftMu, which will not use RaftInterface, (b) call the following
-	// method with the LogSlice, and the callee will make the MsgApp and behave
-	// as if it was sent (i.e., update Next). Since we are not holding
-	// Replica.mu, the callee will need to acquire Replica.mu.
-	MakeMsgAppRaftMuLocked(
-		replicaID roachpb.ReplicaID, start, end uint64, maxSize int64) (raftpb.Message, error)
+	// NB: the [start, end) interval is different from RawNode.LogSlice which
```

Comment:
Why not make it identical? This interface forces writing another wrapper only to change the semantics.